pkg/controller/framework/mutation_detector.go

```diff
@@ -0,0 +1,119 @@
+package framework
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+	"strconv"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/runtime"
+	utildiff "k8s.io/kubernetes/pkg/util/diff"
+)
+
+var mutationDetectionEnabled = false
+
+func init() {
+	mutationDetectionEnabled, _ = strconv.ParseBool(os.Getenv("KUBE_CACHE_MUTATION_DETECTOR"))
+}
+
+type CacheMutationDetector interface {
+	AddObject(obj interface{})
+	Run(stopCh <-chan struct{})
+	// conforms to
+	CompareObjects()
+}
+
+func NewCacheMutationDetector(name string) CacheMutationDetector {
+	if !mutationDetectionEnabled {
+		return dummyMutationDetector{}
+	}
+	return &defaultCacheMutationDetector{name: name, period: 1 * time.Second}
+}
+
+type dummyMutationDetector struct{}
+
+func (dummyMutationDetector) Run(stopCh <-chan struct{}) {
+}
+func (dummyMutationDetector) AddObject(obj interface{}) {
+}
+func (dummyMutationDetector) CompareObjects() {
+}
+
+// cacheMutationDetector gives a way to detect if a cached object has been mutated
+// It has a list of cached objects and their copies.  I haven't thought of a way
+// to see WHO is mutating it, just that its getting mutated
+type defaultCacheMutationDetector struct {
+	name   string
+	period time.Duration
+
+	lock        sync.Mutex
+	cachedObjs  []cacheObj
+	failureFunc func(message string)
+}
+
+// cacheObj holds the actual object and a copy
+type cacheObj struct {
+	cached interface{}
+	copied interface{}
+}
+
+func (d *defaultCacheMutationDetector) Run(stopCh <-chan struct{}) {
+	// we DON'T want protection from panics.  If we're running this code, we want to die
+	go func() {
+		for {
+			d.CompareObjects()
+
+			select {
+			case <-stopCh:
+				return
+			case <-time.After(d.period):
```

Comment:
OK. I will stop objecting.

On Thu, Aug 18, 2016 at 1:08 PM, David Eads <notifications@github.com>
wrote:

> In pkg/controller/framework/mutation_detector.go
> <https://github.com/kubernetes/kubernetes/pull/27784#discussion_r75379572>
> :
>
> > +// cacheObj holds the actual object and a copy
> > +type cacheObj struct {
> > +	cached interface{}
> > +	copied interface{}
> > +}
> > +
> > +func (d *defaultCacheMutationDetector) Run(stopCh <-chan struct{}) {
> > +	// we DON'T want protection from panics.  If we're running this code, we want to die
> > +	go func() {
> > +		for {
> > +			d.CompareObjects()
> > +
> > +			select {
> > +			case <-stopCh:
> > +				return
> > +			case <-time.After(d.period):
>
> I have a hard time convincing myself this is worth it when transient
> changes will probably slip by and even constant changes that happen rarely
> will merely result in flaky behavior.
>
> OTOH I don't have better ideas.
>
> This has caught problems, both in openshift and kube. The problems aren't
> imaginary and they are difficult for reviewers to catch.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27784/files/0cea212f3d60e5ceb403607af009ef1633d7bbfa#r75379572>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngloIYVTdJqYfHbnO9ywwCNWu8BAE5ks5qhLvPgaJpZM4I69qi>
> .
>